## Title: SQL using Python
### Data: Indonesian Movies from IMDB
Dummy data obtained from [Kaggle by dionisiusdh](https://www.kaggle.com/datasets/dionisiusdh/imdb-indonesian-movies)

### Introduction
Welcome to this notebook, where I will engage in introductory SQL operations using SQLite within the Python programming environment.

### Importing Libraries

In [46]:
import pandas as pd
import sqlite3

### Installing iPython SQL

In [47]:
!pip install ipython-sql

### Loading and Connecting Data

**Loading the Data**

In [48]:
df = pd.read_csv("C:/Users/LENOVO/DA_Portfolio/indonesian_movies.csv")

df

title  year  \
0          #FriendButMarried 2  2020   
1                     4 Mantan  2020   
2     Aku Tahu Kapan Kamu Mati  2020   
3                  Anak Garuda  2020   
4                    Dignitate  2020   
...                        ...   ...   
1267     The Tiger from Tjampa  1953   
1268       Enam Djam di Djogja  1951   
1269             Darah dan Doa  1950   
1270         Resia Boroboedoer  1928   
1271        Loetoeng Kasaroeng  1926   

                                            description      genre rating  \
0     Ayudia (Mawar De Jongh) is not satisfied enoug...  Biography    13+   
1     Sara, Airin, Rachel, and Amara were accidental...   Thriller    17+   
2     After apparent death, Siena is able to see sig...     Horror    13+   
3     Good Morning Indonesia, a school for poor orph...  Adventure    13+   
4     Alfi (Al Ghazali) meets Alana (Caitlin Halderm...      Drama    17+   
...                                                 ...        ...    ...   
1267  Set in the 1930s, and narrated like a ballad f...      Drama    NaN   
1268  Depicting the celebrated recapture of the town...      Drama    NaN   
1269  It tells the story of an Indonesian revolution...      Drama    NaN   
1270                                                NaN  Adventure    NaN   
1271                                                NaN    Fantasy    NaN   

      users_rating votes   languages          directors  \
0              6.5   120  Indonesian      Rako Prijanto   
1              6.4     8  Indonesian      Hanny Saputra   
2              5.4    17  Indonesian  Hadrah Daeng Ratu   
3              9.1    27  Indonesian       Faozan Rizal   
4              7.6    33  Indonesian       Fajar Nugros   
...            ...   ...         ...                ...   
1267           6.4    30  Indonesian     D. Djajakusuma   
1268           6.3     9  Indonesian       Usmar Ismail   
1269           6.6    27  Indonesian       Usmar Ismail   
1270           7.0     8  Indonesian                NaN   
1271           7.2    11  Indonesian      L. Heuveldorp   

                                                 actors  runtime  
0     ['Adipati Dolken', 'Mawar Eva de Jongh', 'Vonn...  100 min  
1     ['Ranty Maria', 'Jeff Smith', 'Melanie Berentz...   80 min  
2     ['Natasha Wilona', 'Ria Ricis', 'Al Ghazali', ...   92 min  
3     ['Tissa Biani Azzahra', 'Violla Georgie', 'Aji...  129 min  
4     ['Al Ghazali', 'Caitlin Halderman', 'Giorgino ...  109 min  
...                                                 ...      ...  
1267  ['Wahid Chan', 'Bambang Hermanto', 'R.D. Ismai...   97 min  
1268  ['R.D. Ismail', 'Del Juzar', 'Aedy Moward', 'A...  116 min  
1269  ['Ella Bergen', 'Faridah', 'R.D. Ismail', 'Del...  150 min  
1270  ['Olive Young', nan, nan, nan, nan, nan, nan, ...      NaN  
1271  ['Martoana', 'Oemar', nan, nan, nan, nan, nan,...   60 min  

[1272 rows x 11 columns]

**Connecting the Data to Database**

In [49]:
cnn = sqlite3.connect('sql_python.db')

In [43]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [44]:
%%sql

sqlite:///sql_python.db

### Question 1
**How many movies in the dataset were released after the year 2010?**

In [45]:
%%sql

SELECT COUNT(1) AS movie_after_2010
FROM indonesian_movies
WHERE year > 2010

 * sqlite:///sql_python.db
Done.


[(663,)]

A total of 663 Indonesian films have been released after the year 2010.

### Question 2
**What is the average users' rating for movies in each genre? List the genres in descending order of average rating.**

In [10]:
%%sql

SELECT genre, ROUND(AVG(users_rating), 1) AS avg_users_rating
FROM indonesian_movies
GROUP BY genre
ORDER BY avg_users_rating DESC

 * sqlite:///sql_python.db
Done.


[('History', 7.5),
 ('Biography', 7.4),
 ('Family', 7.1),
 ('Romance', 6.7),
 ('Drama', 6.7),
 ('War', 6.6),
 ('Adventure', 6.3),
 (None, 6.3),
 ('Crime', 6.2),
 ('Thriller', 6.1),
 ('Fantasy', 6.0),
 ('Action', 6.0),
 ('Comedy', 5.9),
 ('Sci-Fi', 5.8),
 ('Animation', 5.8),
 ('Horror', 5.1)]

### Question 3
**Find the top 5 directors (by the number of movies directed) along with the total number of movies they directed.**

In [11]:
%%sql

SELECT directors, COUNT(title) AS num_movies
FROM indonesian_movies
GROUP BY directors
ORDER BY num_movies DESC
LIMIT 5

 * sqlite:///sql_python.db
Done.


[('Nayato Fio Nuala', 61),
 ('Rizal Mantovani', 34),
 ('Hanung Bramantyo', 31),
 ('Arizal', 28),
 ('Sisworo Gautama Putra', 23)]

### Question 4
**Find all the movies that include the actor "Reza Rahadian" in the list of actors. Include the title, year, genre, users' rating, and the rank of each movie's users' rating within the result set, ordered by users' rating in descending order.**

In [39]:
%%sql

SELECT title, year, genre, users_rating,
    RANK() OVER(ORDER BY users_rating DESC) AS rating_rank
FROM indonesian_movies
WHERE actors LIKE '%Reza Rahadian%'
ORDER BY users_rating DESC

 * sqlite:///sql_python.db
Done.


[('Imperfect', 2019, 'Comedy', 7.8, 1),
 ('Kartini', 2017, 'Biography', 7.7, 2),
 ('The Sinking of Van Der Wijck', 2013, 'Drama', 7.6, 3),
 ('Habibie & Ainun', 2012, 'Biography', 7.6, 3),
 ('Hari Untuk Amanda', 2010, 'Drama', 7.6, 3),
 ('The Gift', 2018, 'Drama', 7.5, 6),
 ('Rudy Habibie', 2016, 'Biography', 7.5, 6),
 ('Jakarta Maghrib', 2010, 'Drama', 7.5, 6),
 ('Something in the Way', 2013, 'Drama', 7.4, 9),
 ('Critical Eleven', 2017, 'Drama', 7.3, 10),
 ('Talak 3', 2016, 'Comedy', 7.3, 10),
 ('Guru Bangsa Tjokroaminoto', 2015, 'Biography', 7.3, 10),
 ('Kapan Kawin?', 2015, 'Comedy', 7.3, 10),
 ('Love Story', 2011, 'Drama', 7.3, 10),
 ('My Stupid Boss', 2016, 'Comedy', 7.2, 15),
 ('BrokenHearts', 2012, 'Drama', 7.2, 15),
 ('Dilema', 2011, 'Crime', 7.2, 15),
 ('Hafalan Shalat Delisa', 2011, 'Drama', 7.2, 15),
 ('Alangkah Lucunya (Negeri Ini)', 2010, 'Comedy', 7.2, 15),
 ('Surga Yang Tak Dirindukan 2', 2017, 'Drama', 7.1, 20),
 ('November 10th', 2015, 'Animation', 7.1, 20),
 ('Test Pack', 2012, 'Comedy', 7.1, 20),
 ('Laut Bercermin', 2011, 'Drama', 7.1, 20),
 ('Tanda Tanya', 2011, 'Drama', 7.1, 20),
 ('Habibie & Ainun 3', 2019, 'Biography', 7.0, 25),
 ('Perahu Kertas', 2012, 'Drama', 7.0, 25),
 ('3 Hati, Dua Dunia, Satu Cinta', 2010, 'Comedy', 6.9, 27),
 ('Emak Ingin Naik Haji', 2009, 'Drama', 6.9, 27),
 ('Finding Srimulat', 2013, 'Comedy', 6.8, 29),
 ('Perahu Kertas 2', 2012, 'Drama', 6.8, 29),
 ('Wanita Tetap Wanita', 2013, 'Drama', 6.7, 31),
 ('Perempuan Berkalung Sorban', 2009, 'Drama', 6.7, 31),
 ('My Stupid Boss 2', 2019, 'Adventure', 6.6, 33),
 ('3 Srikandi', 2016, 'Biography', 6.6, 33),
 ('Terpana', 2016, 'Drama', 6.6, 33),
 ('Firegate', 2016, 'Adventure', 6.5, 36),
 ('Toko Barang Mantan', 2020, 'Comedy', 6.3, 37),
 ('Pendekar Tongkat Emas', 2014, 'Action', 6.2, 38),
 ('Kirun + Adul', 2009, 'Comedy', 6.2, 38),
 ('Abracadabra', 2019, 'Comedy', 5.9, 40),
 ('Strawberry Surprise', 2014, 'Drama', 5.9, 40),
 ('Twivortiare', 2019, 'Drama', 5.8, 42),
 ('Film Horor', 2007, 'Comedy', 5.2, 43),
 ('Queen Bee', 2009, 'Drama', 4.6, 44),
 ('Pulau Hantu 2', 2008, 'Horror', 3.8, 45),
 ('Perjaka Terakhir', 2009, 'Drama', 3.6, 46),
 ('Benyamin Biang Kerok', 2018, 'Comedy', 2.9, 47)]

### Question 5
**For each year, find the movie with the highest users' rating. Provide the title, year, and the users' rating for each movie. Additionally, include the rank of each movie's users' rating within its respective year, ordered from highest to lowest rating.**

In [14]:
%%sql

SELECT title, year, users_rating,
        RANK() OVER(PARTITION BY year ORDER BY users_rating DESC) AS rating_rank
FROM indonesian_movies
WHERE (year, users_rating) IN (
        SELECT year, MAX(users_rating)
        FROM indonesian_movies
        GROUP BY year
    )
ORDER BY users_rating DESC

 * sqlite:///sql_python.db
Done.


[('Horas Amang: Tiga Bulan Untuk Selamanya', 2019, 9.4, 1),
 ('Salisiah Adaik', 2013, 9.2, 1),
 ('Jelita Sejuba: Mencintai Kesatria Negara', 2018, 9.1, 1),
 ('Anak Garuda', 2020, 9.1, 1),
 ('Turis Romantis', 2015, 8.9, 1),
 ('Surau dan Silek', 2017, 8.8, 1),
 ('Wage', 2017, 8.8, 1),
 ("Soerabaja '45", 1990, 8.7, 1),
 ('12 Menit', 2014, 8.6, 1),
 ('Sunya', 2016, 8.6, 1),
 ('Mat Peci', 1978, 8.5, 1),
 ('Arini', 1987, 8.5, 1),
 ('Intan Berduri', 1972, 8.4, 1),
 ('Suci Sang Primadona', 1977, 8.4, 1),
 ('Heart 2 Heart', 2010, 8.4, 1),
 ('Roman Picisan', 1980, 8.3, 1),
 ('Titian Serambut Dibelah Tujuh', 1982, 8.2, 1),
 ('Malam Jumat Kliwon', 1986, 8.2, 1),
 ('Sentuhan Rumput yang Bergoyang', 1986, 8.2, 1),
 ('Satu Rumpun?', 2009, 8.2, 1),
 ('Gita Cinta dari SMA', 1979, 8.0, 1),
 ('Permainan Binal', 1995, 8.0, 1),
 ('Syahadat Cinta', 2008, 8.0, 1),
 ('Di Timur Matahari', 2012, 8.0, 1),
 ('Sampai Ujung Dunia', 2012, 8.0, 1),
 ('Ken Arok - Ken Dedes', 1983, 7.9, 1),
 ('Itu Bisa Diatur', 1984, 7.9, 1),
 ('Saur Sepuh 2: Pesanggrahan Keramat', 1989, 7.9, 1),
 ('Si Kabayan dan Anak Jin', 1991, 7.9, 1),
 ('Badut-Badut Kota', 1993, 7.9, 1),
 ('Tentang Cinta', 2007, 7.9, 1),
 ('Asrama Dara', 1958, 7.8, 1),
 ('Serangan Fajar', 1981, 7.8, 1),
 ('Petualangan Sherina', 2000, 7.8, 1),
 ("What's Up with Love?", 2002, 7.8, 1),
 ('The Tarix Jabrix 3', 2011, 7.8, 1),
 ('Kembang Kertas', 1985, 7.7, 1),
 ('Bernapas dalam Lumpur', 1970, 7.6, 1),
 ('Leaf on a Pillow', 1998, 7.6, 1),
 ('Gie', 2005, 7.6, 1),
 ('Janji Joni', 2005, 7.6, 1),
 ('Betina', 2006, 7.6, 1),
 ('Lewat Djam Malam', 1954, 7.5, 1),
 ('Saur Sepuh: Satria Madangkara', 1988, 7.5, 1),
 ('Setetes Noda Manis', 1994, 7.5, 1),
 ('Marsinah', 2001, 7.5, 1),
 ('Tiga Dara', 1956, 7.4, 1),
 ('Cinta Pertama', 1973, 7.4, 1),
 ('Impian Kemarau', 2004, 7.4, 1),
 ('Loetoeng Kasaroeng', 1926, 7.2, 1),
 ('Tarsan Kota', 1974, 7.2, 1),
 ('Bibir Mer', 1992, 7.2, 1),
 ('Arisan!', 2003, 7.2, 1),
 ('Malin Kundang', 1971, 7.1, 1),
 ('Si Pitung Beraksi Kembali', 1976, 7.1, 1),
 ('Resia Boroboedoer', 1928, 7.0, 1),
 ('Lampiasan nafsu', 1996, 7.0, 1),
 ('Tropic of Emerald', 1997, 6.9, 1),
 ('Si Pitung', 1975, 6.8, 1),
 ('Kuldesak', 1999, 6.8, 1),
 ('Darah dan Doa', 1950, 6.6, 1),
 ('The Tiger from Tjampa', 1953, 6.4, 1),
 ('Enam Djam di Djogja', 1951, 6.3, 1),
 ('Badai Selatan', 1962, 6.3, 1),
 ('Pedjuang', 1961, 6.2, 1),
 ('Tamu Agung', 1955, 5.0, 1)]